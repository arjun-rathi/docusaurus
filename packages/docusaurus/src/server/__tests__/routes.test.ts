/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import loadRoutes from '../routes';
import type {RouteConfig} from '@docusaurus/types';

describe('loadRoutes', () => {
  it('loads nested route config', async () => {
    const nestedRouteConfig: RouteConfig = {
      component: '@theme/DocPage',
      path: '/docs:route',
      modules: {
        docsMetadata: 'docs-b5f.json',
      },
      routes: [
        {
          path: '/docs/hello',
          component: '@theme/DocItem',
          exact: true,
          modules: {
            content: 'docs/hello.md',
            metadata: 'docs-hello-da2.json',
          },
          sidebar: 'main',
        },
        {
          path: 'docs/foo/baz',
          component: '@theme/DocItem',
          modules: {
            content: 'docs/foo/baz.md',
            metadata: 'docs-foo-baz-dd9.json',
          },
          sidebar: 'secondary',
          'key:a': 'containing colon',
          "key'b": 'containing quote',
          'key"c': 'containing double quote',
          'key,d': 'containing comma',
          字段: 'containing unicode',
        },
      ],
    };
    await expect(
      loadRoutes([nestedRouteConfig], '/'),
    ).resolves.toMatchSnapshot();
  });

  it('loads flat route config', async () => {
    const flatRouteConfig: RouteConfig = {
      path: '/blog',
      component: '@theme/BlogListPage',
      exact: true,
      modules: {
        items: [
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
              query: {
                truncated: true,
              },
            },
            metadata: 'blog-2018-12-14-happy-first-birthday-slash-d2c.json',
          },
          {
            content: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            metadata: null,
          },
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            },
            metadata: null,
          },
        ],
      },
    };
    await expect(loadRoutes([flatRouteConfig], '/')).resolves.toMatchSnapshot();
  });

  it('rejects invalid route config', async () => {
    const routeConfigWithoutPath = {
      component: 'hello/world.js',
    } as RouteConfig;

    await expect(loadRoutes([routeConfigWithoutPath], '/')).rejects
      .toThrowErrorMatchingInlineSnapshot(`
            "Invalid route config: path must be a string and component is required.
            {\\"component\\":\\"hello/world.js\\"}"
          `);

    const routeConfigWithoutComponent = {
      path: '/hello/world',
    } as RouteConfig;

    await expect(loadRoutes([routeConfigWithoutComponent], '/')).rejects
      .toThrowErrorMatchingInlineSnapshot(`
            "Invalid route config: path must be a string and component is required.
            {\\"path\\":\\"/hello/world\\"}"
          `);
  });

  it('loads route config with empty (but valid) path string', async () => {
    const routeConfig = {
      path: '',
      component: 'hello/world.js',
    } as RouteConfig;

    await expect(loadRoutes([routeConfig], '/')).resolves.toMatchSnapshot();
  });
});
