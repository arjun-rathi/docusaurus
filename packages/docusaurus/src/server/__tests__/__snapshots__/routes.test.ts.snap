// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadRoutes loads flat route config 1`] = `
{
  "registry": {
    "component---theme-blog-list-pagea-6-a-7ba": {
      "loader": "() => import(/* webpackChunkName: 'component---theme-blog-list-pagea-6-a-7ba' */ '@theme/BlogListPage')",
      "modulePath": "@theme/BlogListPage",
    },
    "content---blog-0-b-4-09e": {
      "loader": "() => import(/* webpackChunkName: 'content---blog-0-b-4-09e' */ 'blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true')",
      "modulePath": "blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true",
    },
    "content---blog-7-b-8-fd9": {
      "loader": "() => import(/* webpackChunkName: 'content---blog-7-b-8-fd9' */ 'blog/2018-12-14-Happy-First-Birthday-Slash.md')",
      "modulePath": "blog/2018-12-14-Happy-First-Birthday-Slash.md",
    },
    "metadata---blog-0-b-6-74c": {
      "loader": "() => import(/* webpackChunkName: 'metadata---blog-0-b-6-74c' */ 'blog-2018-12-14-happy-first-birthday-slash-d2c.json')",
      "modulePath": "blog-2018-12-14-happy-first-birthday-slash-d2c.json",
    },
  },
  "routesChunkNames": {
    "/blog-1e7": {
      "component": "component---theme-blog-list-pagea-6-a-7ba",
      "items": [
        {
          "content": "content---blog-0-b-4-09e",
          "metadata": "metadata---blog-0-b-6-74c",
        },
        {
          "content": "content---blog-7-b-8-fd9",
          "metadata": null,
        },
        {
          "content": "content---blog-7-b-8-fd9",
          "metadata": null,
        },
      ],
    },
  },
  "routesConfig": "
import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/blog',
    component: ComponentCreator('/blog','1e7'),
    exact: true
  },
  {
    path: '*',
    component: ComponentCreator('*')
  }
];
",
  "routesPaths": [
    "/404.html",
    "/blog",
  ],
}
`;

exports[`loadRoutes loads nested route config 1`] = `
{
  "registry": {
    "component---theme-doc-item-178-a40": {
      "loader": "() => import(/* webpackChunkName: 'component---theme-doc-item-178-a40' */ '@theme/DocItem')",
      "modulePath": "@theme/DocItem",
    },
    "component---theme-doc-page-1-be-9be": {
      "loader": "() => import(/* webpackChunkName: 'component---theme-doc-page-1-be-9be' */ '@theme/DocPage')",
      "modulePath": "@theme/DocPage",
    },
    "content---docs-foo-baz-8-ce-61e": {
      "loader": "() => import(/* webpackChunkName: 'content---docs-foo-baz-8-ce-61e' */ 'docs/foo/baz.md')",
      "modulePath": "docs/foo/baz.md",
    },
    "content---docs-helloaff-811": {
      "loader": "() => import(/* webpackChunkName: 'content---docs-helloaff-811' */ 'docs/hello.md')",
      "modulePath": "docs/hello.md",
    },
    "docsMetadata---docs-routef-34-881": {
      "loader": "() => import(/* webpackChunkName: 'docsMetadata---docs-routef-34-881' */ 'docs-b5f.json')",
      "modulePath": "docs-b5f.json",
    },
    "metadata---docs-foo-baz-2-cf-fa7": {
      "loader": "() => import(/* webpackChunkName: 'metadata---docs-foo-baz-2-cf-fa7' */ 'docs-foo-baz-dd9.json')",
      "modulePath": "docs-foo-baz-dd9.json",
    },
    "metadata---docs-hello-956-741": {
      "loader": "() => import(/* webpackChunkName: 'metadata---docs-hello-956-741' */ 'docs-hello-da2.json')",
      "modulePath": "docs-hello-da2.json",
    },
  },
  "routesChunkNames": {
    "/docs/hello-44b": {
      "component": "component---theme-doc-item-178-a40",
      "content": "content---docs-helloaff-811",
      "metadata": "metadata---docs-hello-956-741",
    },
    "/docs:route-52d": {
      "component": "component---theme-doc-page-1-be-9be",
      "docsMetadata": "docsMetadata---docs-routef-34-881",
    },
    "docs/foo/baz-070": {
      "component": "component---theme-doc-item-178-a40",
      "content": "content---docs-foo-baz-8-ce-61e",
      "metadata": "metadata---docs-foo-baz-2-cf-fa7",
    },
  },
  "routesConfig": "
import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/docs:route',
    component: ComponentCreator('/docs:route','52d'),
    routes: [
      {
        path: '/docs/hello',
        component: ComponentCreator('/docs/hello','44b'),
        exact: true,
        sidebar: \\"main\\"
      },
      {
        path: 'docs/foo/baz',
        component: ComponentCreator('docs/foo/baz','070'),
        sidebar: \\"secondary\\",
        \\"key:a\\": \\"containing colon\\",
        \\"key'b\\": \\"containing quote\\",
        \\"key\\\\\\"c\\": \\"containing double quote\\",
        \\"key,d\\": \\"containing comma\\",
        字段: \\"containing unicode\\"
      }
    ]
  },
  {
    path: '*',
    component: ComponentCreator('*')
  }
];
",
  "routesPaths": [
    "/404.html",
    "/docs/hello",
    "docs/foo/baz",
  ],
}
`;

exports[`loadRoutes loads route config with empty (but valid) path string 1`] = `
{
  "registry": {
    "component---hello-world-jse-0-f-b6c": {
      "loader": "() => import(/* webpackChunkName: 'component---hello-world-jse-0-f-b6c' */ 'hello/world.js')",
      "modulePath": "hello/world.js",
    },
  },
  "routesChunkNames": {
    "-b2a": {
      "component": "component---hello-world-jse-0-f-b6c",
    },
  },
  "routesConfig": "
import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '',
    component: ComponentCreator('','b2a')
  },
  {
    path: '*',
    component: ComponentCreator('*')
  }
];
",
  "routesPaths": [
    "/404.html",
    "",
  ],
}
`;
